File: Leetcode/Python/_343.py
class Solution:
    def integerBreak(self, n: int) -> int:
        if n <= 3:
            return n - 1

        dp = [0] * (n + 1)

        for i in [1, 2, 3]:
            dp[i] = i

        for n in range(4, n + 1):
            output = n
            for i in range(2, n):
                output = max(output, i * dp[n - i])

            dp[n] = output

        return dp[n]

File: Leetcode/cpp/343.cpp
class Solution {
public:
    int integerBreak(int n) {
        if (n <= 3) {
            return n - 1;
        }
        vector<int> dp(n + 1);

        for (int i = 0; i < 4; i++) {
            dp[i] = i;
        }

        for (int j = 4; j <= n; j++) {
            int ans = j;
            for (int i = 0; i < j; i++) {
                ans = std::max(ans, i * dp[j - i]);
            }
            dp[j] = ans;
        }
        return dp[n];
    }
};
